server/src/main/java/org/elasticsearch/repositories/blobstore/ChunkedBlobOutputStream.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.repositories.blobstore;
+
+import org.elasticsearch.common.io.stream.ReleasableBytesStreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.core.Releasables;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Base class for doing chunked writes to a blob store. Some blob stores require either up-front knowledge of the size of the blob that
+ * will be written or writing it in chunks that are then joined into the final blob at the end of the write. This class provides a basis
+ * on which to implement an output stream that encapsulates such a chunked write.
+ *
+ * @param <T> type of chunk identifier
+ */
+public abstract class ChunkedBlobOutputStream<T> extends OutputStream {
+
+    /**
+     * List of identifiers of already written chunks.
+     */
+    protected final List<T> parts = new ArrayList<>();
+
+    /**
+     * Size of the write buffer above which it must be flushed to storage.
+     */
+    private final long maxBytesToBuffer;
+
+    /**
+     * Big arrays to be able to allocate buffers from pooled bytes.
+     */
+    private final BigArrays bigArrays;
+
+    /**
+     * Current write buffer.
+     */
+    protected ReleasableBytesStreamOutput buffer;
+
+    /**
+     * Set to true once no more calls to {@link #write} are expected and the blob has been received by {@link #write} in full so that
+     * {@link #close()} knows whether to clean up existing chunks or finish a chunked write.
+     */
+    protected boolean successful = false;
+
+    /**
+     * Number of bytes written to storage writeBlso far.
+     */
+    protected long written = 0L;
+
+    protected ChunkedBlobOutputStream(BigArrays bigArrays, long maxBytesToBuffer) {
+        this.bigArrays = bigArrays;
+        this.maxBytesToBuffer = maxBytesToBuffer;
+        buffer = new ReleasableBytesStreamOutput(bigArrays);
+    }
+
+    @Override
+    public final void write(int b) throws IOException {
+        buffer.write(b);
+        maybeFlushBuffer();
+    }
+
+    @Override
+    public final void write(byte[] b, int off, int len) throws IOException {
+        buffer.write(b, off, len);
+        maybeFlushBuffer();
```

Comment:
I think it is possible to exceed the buffer size since maybeFlushBuffer() checks that the buffer size > max bytes to buffer _after_ the bytes are written to the ReleasableBytesStreamOutput

Reply:
That's right. I figured this wasn't a big deal in the real world though as write buffers are probably reasonably small in all cases. We could technically be a little smarter here and not copy to the internal buffer for the last write before a flush and instead concatenate the `BytesReference`. But if you think about it, the internal buffer is made up of 16k pages and we're working with pooled and O(MB) buffers here, is it really worth the relatively large complexity increase for the relatively minor memory + copy saving? :)